File: SpeedTest/implementations/zig/mergesort.zig
const std = @import("std");
const time = std.time;

fn merge(arr: []i32, left: []i32, right: []i32) void {
    var i: usize = 0;
    var j: usize = 0;
    var k: usize = 0;

    while (i < left.len and j < right.len) {
        if (left[i] <= right[j]) {
            arr[k] = left[i];
            i += 1;
        } else {
            arr[k] = right[j];
            j += 1;
        }
        k += 1;
    }

    while (i < left.len) {
        arr[k] = left[i];
        i += 1;
        k += 1;
    }

    while (j < right.len) {
        arr[k] = right[j];
        j += 1;
        k += 1;
    }
}

fn mergeSort(arr: []i32) void {
    if (arr.len <= 1) {
        return;
    }

    const mid = arr.len / 2;
    const left = arr[0..mid];
    const right = arr[mid..];

    mergeSort(left);
    mergeSort(right);

    var temp = std.heap.page_allocator.alloc(i32, arr.len) catch unreachable;
    defer std.heap.page_allocator.free(temp);

    std.mem.copy(i32, temp[0..], left);
    std.mem.copy(i32, temp[mid..], right);

    merge(arr, temp[0..mid], temp[mid..]);
}

fn generateRandomArray(allocator: std.mem.Allocator, size: usize) ![]i32 {
    const arr = try allocator.alloc(i32, size);
    var rng = std.crypto.random;

    for (arr) |*item| {
        item.* = @intCast(rng.intRangeAtMost(i32, 0, 9999));
    }

    return arr;
}

pub fn main() void {
    const ARRAY_SIZE = 1000;
    const ITERATIONS = 100;

    var i: usize = 0;
    while (i < ITERATIONS) : (i += 1) {
        const arr = generateRandomArray(std.heap.page_allocator, ARRAY_SIZE) catch unreachable;
        defer std.heap.page_allocator.free(arr);

        mergeSort(arr);
    }
}

File: SpeedTest/implementations/perl/mergesort.pl
#!/usr/bin/env perl
use strict;
use warnings;

sub merge {
    my ($arr, $left, $right) = @_;
    my ($i, $j, $k) = (0, 0, 0);

    while ($i < @$left && $j < @$right) {
        if ($left->[$i] <= $right->[$j]) {
            $arr->[$k] = $left->[$i];
            $i++;
        } else {
            $arr->[$k] = $right->[$j];
            $j++;
        }
        $k++;
    }

    while ($i < @$left) {
        $arr->[$k] = $left->[$i];
        $i++;
        $k++;
    }

    while ($j < @$right) {
        $arr->[$k] = $right->[$j];
        $j++;
        $k++;
    }
}

sub mergeSort {
    my ($arr) = @_;

    if (@$arr <= 1) {
        return;
    }

    my $mid = int(@$arr / 2);
    my @left = @$arr[0 .. $mid - 1];
    my @right = @$arr[$mid .. @$arr - 1];

    mergeSort(\@left);
    mergeSort(\@right);

    merge($arr, \@left, \@right);
}

sub generateRandomArray {
    my ($size) = @_;
    return [map { int(rand(10000)) } 1 .. $size];
}

my $ARRAY_SIZE = 1000;
my $ITERATIONS = 100;

for (1 .. $ITERATIONS) {
    my $arr = generateRandomArray($ARRAY_SIZE);
    mergeSort($arr);
}
